syntax = "proto3";

package api.order;

option go_package = "OrderService/api/order;order";
option java_multiple_files = true;
option java_package = "api.order";

service OrderService {
	rpc PlaceOrder(PlaceOrderReq) returns (PlaceOrderResp) {}
	rpc ListOrder(ListOrderReq) returns (ListOrderResp) {}
	rpc GetOrderById(GetOrderByIdReq) returns (GetOrderByIdResp) {}
	rpc	DelOrderById(DelOrderByIdReq) returns (DelOrderByIdResp) {}
	rpc MarkOrderPaid(MarkOrderPaidReq) returns (MarkOrderPaidResp) {}
	rpc MarkOrderCancel(MarkOrderCancelReq) returns (MarkOrderCancelResp) {}
}

// PlaceOrderReq
message Address {
	string street_address = 1;
	string city = 2;
	string country = 3;
	uint32 zip_code = 4;
}

message OrderItem {
	int64 product_id = 1;
	uint32 quantity = 2;
	float cost = 3;
}

message PlaceOrderReq {
	int64 user_id = 1;
	Address address = 3;
	string phone_number = 4;
	repeated OrderItem order_items = 5;
}
//PlaceOrderResp

message PlaceOrderResp {
	int64 order_id = 1;
}

//

message ListOrderReq {
	int64 user_id = 1;
}
// ListOrderResp
message Order {
	repeated OrderItem order_items = 1;
	int64 order_id = 2;
	int64 user_id = 3;
	float OriginalCharge = 4;
	float ActualPayment = 5;
	bool IsFreeShipping = 6;
	float ShippingFee = 7;
	Address address = 8;
	string phone_number = 9;
	string order_state = 10;
}

message ListOrderResp {
	int64 total = 1;
	repeated Order orders = 2;
}

// GetOrderById
message GetOrderByIdReq {
	int64 user_id = 1;
	int64 order_id = 2;
}

message GetOrderByIdResp {
	Order order = 1;
}

// DelOrderById
message DelOrderByIdReq {
	int64 user_id = 1;
	int64 order_id = 2;
}

message DelOrderByIdResp {
	bool state=1;
}

// MarkOrderPaidReq
message MarkOrderPaidReq {
	int64 order_id = 1;
}

message MarkOrderPaidResp {
	bool state = 1;
}


// MarkOrderCancelReq
message MarkOrderCancelReq {
	int64 user_id = 1;
	int64 order_id = 2;
}

message MarkOrderCancelResp {
	bool state = 1;
}

